bee-inx/src/ledger.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use bee_block as bee;
+use inx::proto;
+
+use crate::Raw;
+
+/// Represents a new output in the ledger.
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LedgerOutput {
+    pub output_id: bee::output::OutputId,
+    pub block_id: bee::BlockId,
+    pub milestone_index_booked: u32,
+    pub milestone_timestamp_booked: u32,
+    pub output: Raw<bee::output::Output>,
+}
+
+/// Represents a spent output in the ledger.
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LedgerSpent {
+    pub output: LedgerOutput,
+    pub transaction_id_spent: bee::payload::transaction::TransactionId,
+    pub milestone_index_spent: u32,
+    pub milestone_timestamp_spent: u32,
+}
+
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct UnspentOutput {
+    pub ledger_index: u32,
+    pub output: LedgerOutput,
+}
+
+/// Represents an update to ledger.
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LedgerUpdate {
+    pub milestone_index: u32,
+    pub created: Box<[LedgerOutput]>,
+    pub consumed: Box<[LedgerSpent]>,
+}
+
+impl TryFrom<proto::LedgerOutput> for LedgerOutput {
+    type Error = bee::InxError;
+
+    fn try_from(value: proto::LedgerOutput) -> Result<Self, Self::Error> {
+        Ok(Self {
+            output_id: value
+                .output_id
+                .ok_or(Self::Error::MissingField("output_id"))?
+                .try_into()?,
+            block_id: value
+                .block_id
+                .ok_or(Self::Error::MissingField("message_id"))?
```

Comment:
```suggestion
                .ok_or(Self::Error::MissingField("block_id"))?
```

Reply:
Created a convenience macro that prevents these type of mistakes.